src/clj/ataru/applications/application_store.clj

```diff
@@ -370,6 +376,43 @@
                           :operation        audit-log/operation-new
                           :organization-oid organization-oid}))))))
 
+(defn get-application-hakukohde-reviews
+  [application-key]
+  (mapv ->kebab-case-kw (exec-db :db yesql-get-application-hakukohde-reviews {:application_key application-key})))
+
+(defn save-application-hakukohde-review
+  [virkailija application-key hakukohde hakukohde-review-requirement hakukohde-review-state session]
+  (jdbc/with-db-transaction [conn {:datasource (db/get-datasource :db)}]
+                            (let [connection                  {:connection conn}
+                                  review-to-store             {:application_key application-key
+                                                               :requirement     hakukohde-review-requirement
+                                                               :state           hakukohde-review-state
+                                                               :hakukohde       hakukohde}
+                                  existing-duplicate-review   (yesql-get-existing-application-review review-to-store connection)
+                                  existing-requirement-review (yesql-get-existing-requirement-review review-to-store connection)
+                                  username                    (get-in session [:identity :username])
+                                  organization-oid            (get-in session [:identity :organizations 0 :oid])]
+                              (when (empty? existing-duplicate-review)
```

Comment:
Does this mean that upsert is never used as upsert?

Reply:
No, this catches cases where the application, hakukohde, review-requirement and review-state are all the same (ie. no update needs to be done as nothing would change). The upsert is for cases where the review-state changes while application, hakukohde and review-requirement are the same. At least that is the plan.